File: ts/shared/vectors.test.ts
/**
 * Vector tests for the TypeScript shared library.
 * Loads the same canonical fixtures as the Go tests and asserts exact byte output.
 * A failure here before any application code is debugged points to a specific
 * serialization layer disagreement with the Go implementation.
 */
import { readFileSync } from "fs";
import { join } from "path";
import { createHash } from "crypto";
import { hashLeaf, hashNode, entryHash, computeRoot, inclusionProof, verifyInclusion } from "./merkle.js";
import { encodeDataAssertion, encodeNullEntry } from "./cbor.js";
import { checkpointBody } from "./checkpoint.js";

const VECTORS_PATH = join(import.meta.dirname ?? __dirname, "../../test-vectors/vectors.json");

interface Vector {
  id: string;
  description: string;
  input: unknown;
  expected: unknown;
}

function loadVectors(): Map<string, Vector> {
  const raw = JSON.parse(readFileSync(VECTORS_PATH, "utf8")) as { vectors: Vector[] };
  const m = new Map<string, Vector>();
  for (const v of raw.vectors) m.set(v.id, v);
  return m;
}

function hex(b: Uint8Array): string {
  return Buffer.from(b).toString("hex");
}

function fromHex(s: string): Uint8Array {
  return new Uint8Array(Buffer.from(s, "hex"));
}

let passed = 0;
let failed = 0;

function test(name: string, fn: () => void): void {
  try {
    fn();
    console.log(`  ✓ ${name}`);
    passed++;
  } catch (e: unknown) {
    const msg = e instanceof Error ? e.message : String(e);
    console.log(`  ✗ ${name}\n      ${msg}`);
    failed++;
  }
}

function assert(got: string, want: string, label: string): void {
  if (got !== want) {
    throw new Error(`${label}:\n      got  ${got}\n      want ${want}`);
  }
}

// --- Run tests ---

const vs = loadVectors();
console.log("\nMTA-QR TypeScript vector tests\n");

// Vector 1: checkpoint-body-v1
test("checkpoint-body-v1", () => {
  const v = vs.get("checkpoint-body-v1")!;
  const input = v.input as { origin: string; tree_size: number; root_hash_hex: string };
  const expected = v.expected as { checkpoint_body_hex: string; byte_length: number };

  const rootHash = fromHex(input.root_hash_hex);
  const body = checkpointBody(input.origin, BigInt(input.tree_size), rootHash);

  assert(String(body.length), String(expected.byte_length), "body length");
  assert(hex(body), expected.checkpoint_body_hex, "body hex");

  if (body[body.length - 1] !== 0x0a) {
    throw new Error("body must end with \\n (0x0a)");
  }
});

// Vector 2: null-entry-hash
test("null-entry-hash", () => {
  const v = vs.get("null-entry-hash")!;
  const expected = v.expected as { entry_hash_hex: string };

  const tbs = encodeNullEntry();
  assert(hex(tbs), "00", "null entry TBS");

  const hash = entryHash(tbs);
  assert(hex(hash), expected.entry_hash_hex, "null entry hash");
});

// Vector 3: data-assertion-cbor
test("data-assertion-cbor", () => {
  const v = vs.get("data-assertion-cbor")!;
  const input = v.input as {
    issuance_time: number;
    expiry_time: number;
    schema_id: number;
    claims: Record<string, string>;
  };
  const expected = v.expected as { tbs_hex: string; entry_hash_hex: string };

  const tbs = encodeDataAssertion({
    times: [input.issuance_time, input.expiry_time],
    schemaId: input.schema_id,
    claims: input.claims,
  });

  assert(hex(tbs), expected.tbs_hex, "TBS hex");

  const hash = entryHash(tbs);
  assert(hex(hash), expected.entry_hash_hex, "entry_hash");
});

// Vector 4: merkle-four-entry-tree
test("merkle-four-entry-tree", () => {
  const v = vs.get("merkle-four-entry-tree")!;
  const input = v.input as { leaves: { label: string; data_hex: string }[] };
  const expected = v.expected as {
    leaf_hashes: string[];
    internal_nodes: { H01: string; H23: string };
    root: string;
    inclusion_proof_index2: { entry_index: number; tree_size: number; proof: string[] };
  };

  const leaves = input.leaves.map((l) => hashLeaf(fromHex(l.data_hex)));

  for (let i = 0; i < leaves.length; i++) {
    assert(hex(leaves[i]), expected.leaf_hashes[i], `leaf[${i}]`);
  }

  const H01 = hashNode(leaves[0], leaves[1]);
  const H23 = hashNode(leaves[2], leaves[3]);
  assert(hex(H01), expected.internal_nodes.H01, "H01");
  assert(hex(H23), expected.internal_nodes.H23, "H23");

  const root = computeRoot(leaves);
  assert(hex(root), expected.root, "root");

  const ip = expected.inclusion_proof_index2;
  const proof = inclusionProof(leaves, ip.entry_index, ip.tree_size);
  if (proof.length !== ip.proof.length) {
    throw new Error(`proof length: got ${proof.length}, want ${ip.proof.length}`);
  }
  for (let i = 0; i < proof.length; i++) {
    assert(hex(proof[i]), ip.proof[i], `proof[${i}]`);
  }

  // Round-trip: verify the proof.
  verifyInclusion(leaves[2], ip.entry_index, ip.tree_size, proof, root);
});

// Vector 5: entry-hash-construction
test("entry-hash-construction", () => {
  const v = vs.get("entry-hash-construction")!;
  const input = v.input as { tbs_hex: string };
  const expected = v.expected as { preimage_hex: string; entry_hash_hex: string };

  const tbs = fromHex(input.tbs_hex);

  const preimage = new Uint8Array(1 + tbs.length);
  preimage[0] = 0x00;
  preimage.set(tbs, 1);
  assert(hex(preimage), expected.preimage_hex, "preimage");

  const h = createHash("sha256").update(preimage).digest();
  assert(h.toString("hex"), expected.entry_hash_hex, "entry_hash via raw SHA-256");

  // Also via entryHash helper.
  assert(hex(entryHash(tbs)), expected.entry_hash_hex, "entry_hash via helper");
});

// --- Summary ---
console.log(`\n${passed + failed} tests: ${passed} passed, ${failed} failed\n`);
if (failed > 0) process.exit(1);

File: go/shared/vectors_test.go
package vectors_test

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"os"
	"testing"

	"fmt"

	"github.com/mta-qr/demo/shared/cascade"
	mtacbor "github.com/mta-qr/demo/shared/cbor"
	"github.com/mta-qr/demo/shared/checkpoint"
	"github.com/mta-qr/demo/shared/merkle"
	"github.com/mta-qr/demo/shared/payload"
)

// vectorFile is the path to the canonical test vectors JSON, relative to
// the repo root. Adjust if running from a different working directory.
const vectorFile = "../../test-vectors/vectors.json"

type vectors struct {
	Vectors []vector `json:"vectors"`
}

type vector struct {
	ID          string          `json:"id"`
	Description string          `json:"description"`
	Input       json.RawMessage `json:"input"`
	Expected    json.RawMessage `json:"expected"`
}

func loadVectors(t *testing.T) map[string]vector {
	t.Helper()
	data, err := os.ReadFile(vectorFile)
	if err != nil {
		t.Fatalf("load vectors: %v", err)
	}
	var vs vectors
	if err := json.Unmarshal(data, &vs); err != nil {
		t.Fatalf("parse vectors: %v", err)
	}
	m := make(map[string]vector, len(vs.Vectors))
	for _, v := range vs.Vectors {
		m[v.ID] = v
	}
	return m
}

func mustDecodeHex(t *testing.T, s string) []byte {
	t.Helper()
	b, err := hex.DecodeString(s)
	if err != nil {
		t.Fatalf("hex decode %q: %v", s, err)
	}
	return b
}

// --- Vector: checkpoint-body-v1 ---

func TestCheckpointBody(t *testing.T) {
	vs := loadVectors(t)
	v := vs["checkpoint-body-v1"]

	var input struct {
		Origin      string `json:"origin"`
		TreeSize    uint64 `json:"tree_size"`
		RootHashHex string `json:"root_hash_hex"`
	}
	if err := json.Unmarshal(v.Input, &input); err != nil {
		t.Fatalf("parse input: %v", err)
	}
	var expected struct {
		CheckpointBodyHex string `json:"checkpoint_body_hex"`
		ByteLength        int    `json:"byte_length"`
	}
	if err := json.Unmarshal(v.Expected, &expected); err != nil {
		t.Fatalf("parse expected: %v", err)
	}

	rootHash := mustDecodeHex(t, input.RootHashHex)
	body := checkpoint.Body(input.Origin, input.TreeSize, rootHash)

	if len(body) != expected.ByteLength {
		t.Errorf("body length: got %d, want %d", len(body), expected.ByteLength)
	}

	gotHex := hex.EncodeToString(body)
	if gotHex != expected.CheckpointBodyHex {
		t.Errorf("checkpoint body mismatch:\ngot  %s\nwant %s", gotHex, expected.CheckpointBodyHex)
	}

	// Verify the body ends with \n (not just the whole string but literally the last byte).
	if body[len(body)-1] != '\n' {
		t.Errorf("checkpoint body must end with \\n; last byte is 0x%02x", body[len(body)-1])
	}
}

// --- Vector: null-entry-hash ---

func TestNullEntryHash(t *testing.T) {
	vs := loadVectors(t)
	v := vs["null-entry-hash"]

	var expected struct {
		EntryHashHex string `json:"entry_hash_hex"`
	}
	if err := json.Unmarshal(v.Expected, &expected); err != nil {
		t.Fatalf("parse expected: %v", err)
	}

	// null_entry TBS is a single 0x00 byte.
	tbs := []byte{0x00}
	hash := merkle.EntryHash(tbs)
	gotHex := hex.EncodeToString(hash)

	if gotHex != expected.EntryHashHex {
		t.Errorf("null entry hash mismatch:\ngot  %s\nwant %s", gotHex, expected.EntryHashHex)
	}
}

// --- Vector: data-assertion-cbor ---

func TestDataAssertionCBOR(t *testing.T) {
	vs := loadVectors(t)
	v := vs["data-assertion-cbor"]

	var input struct {
		IssuanceTime uint64         `json:"issuance_time"`
		ExpiryTime   uint64         `json:"expiry_time"`
		SchemaID     uint64         `json:"schema_id"`
		Claims       map[string]any `json:"claims"`
	}
	if err := json.Unmarshal(v.Input, &input); err != nil {
		t.Fatalf("parse input: %v", err)
	}
	var expected struct {
		TBSHex        string `json:"tbs_hex"`
		EntryHashHex  string `json:"entry_hash_hex"`
	}
	if err := json.Unmarshal(v.Expected, &expected); err != nil {
		t.Fatalf("parse expected: %v", err)
	}

	tbs, err := mtacbor.EncodeDataAssertion(
		input.IssuanceTime,
		input.ExpiryTime,
		input.SchemaID,
		input.Claims,
	)
	if err != nil {
		t.Fatalf("EncodeDataAssertion: %v", err)
	}

	gotTBSHex := hex.EncodeToString(tbs)
	if gotTBSHex != expected.TBSHex {
		t.Errorf("TBS mismatch:\ngot  %s\nwant %s", gotTBSHex, expected.TBSHex)
	}

	entryHash := merkle.EntryHash(tbs)
	gotHashHex := hex.EncodeToString(entryHash)
	if gotHashHex != expected.EntryHashHex {
		t.Errorf("entry_hash mismatch:\ngot  %s\nwant %s", gotHashHex, expected.EntryHashHex)
	}

	// Canonicalization round-trip check.
	if err := mtacbor.RoundTripCanonical(tbs[1:]); err != nil {
		t.Errorf("CBOR round-trip canonical check failed: %v", err)
	}
}

// --- Vector: merkle-four-entry-tree ---

func TestMerkleFourEntryTree(t *testing.T) {
	vs := loadVectors(t)
	v := vs["merkle-four-entry-tree"]

	var input struct {
		Leaves []struct {
			Label   string `json:"label"`
			DataHex string `json:"data_hex"`
		} `json:"leaves"`
	}
	if err := json.Unmarshal(v.Input, &input); err != nil {
		t.Fatalf("parse input: %v", err)
	}
	var expected struct {
		LeafHashes []string `json:"leaf_hashes"`
		InternalNodes struct {
			H01 string `json:"H01"`
			H23 string `json:"H23"`
		} `json:"internal_nodes"`
		Root                 string `json:"root"`
		InclusionProofIndex2 struct {
			EntryIndex int      `json:"entry_index"`
			TreeSize   int      `json:"tree_size"`
			Proof      []string `json:"proof"`
		} `json:"inclusion_proof_index2"`
	}
	if err := json.Unmarshal(v.Expected, &expected); err != nil {
		t.Fatalf("parse expected: %v", err)
	}

	// Compute leaf hashes using SHA-256(0x00 || data).
	leaves := make([][]byte, len(input.Leaves))
	for i, l := range input.Leaves {
		data := mustDecodeHex(t, l.DataHex)
		leaves[i] = merkle.HashLeaf(data)
		gotHex := hex.EncodeToString(leaves[i])
		if gotHex != expected.LeafHashes[i] {
			t.Errorf("leaf[%d] hash mismatch:\ngot  %s\nwant %s", i, gotHex, expected.LeafHashes[i])
		}
	}

	// Compute internal nodes.
	H01 := merkle.HashNode(leaves[0], leaves[1])
	H23 := merkle.HashNode(leaves[2], leaves[3])
	gotH01 := hex.EncodeToString(H01)
	gotH23 := hex.EncodeToString(H23)
	if gotH01 != expected.InternalNodes.H01 {
		t.Errorf("H01 mismatch:\ngot  %s\nwant %s", gotH01, expected.InternalNodes.H01)
	}
	if gotH23 != expected.InternalNodes.H23 {
		t.Errorf("H23 mismatch:\ngot  %s\nwant %s", gotH23, expected.InternalNodes.H23)
	}

	// Compute root.
	root, err := merkle.Root(leaves)
	if err != nil {
		t.Fatalf("Root: %v", err)
	}
	gotRoot := hex.EncodeToString(root)
	if gotRoot != expected.Root {
		t.Errorf("root mismatch:\ngot  %s\nwant %s", gotRoot, expected.Root)
	}

	// Validate inclusion proof for entry_index 2.
	ip := expected.InclusionProofIndex2
	proof, err := merkle.InclusionProof(leaves, ip.EntryIndex, ip.TreeSize)
	if err != nil {
		t.Fatalf("InclusionProof: %v", err)
	}
	if len(proof) != len(ip.Proof) {
		t.Fatalf("proof length: got %d, want %d", len(proof), len(ip.Proof))
	}
	for i, h := range proof {
		gotHex := hex.EncodeToString(h)
		if gotHex != ip.Proof[i] {
			t.Errorf("proof[%d] mismatch:\ngot  %s\nwant %s", i, gotHex, ip.Proof[i])
		}
	}

	// Verify the proof round-trips.
	if err := merkle.VerifyInclusion(leaves[2], ip.EntryIndex, ip.TreeSize, proof, root); err != nil {
		t.Errorf("VerifyInclusion failed: %v", err)
	}
}

// --- Vector: entry-hash-construction ---

func TestEntryHashConstruction(t *testing.T) {
	vs := loadVectors(t)
	v := vs["entry-hash-construction"]

	var input struct {
		TBSHex string `json:"tbs_hex"`
	}
	if err := json.Unmarshal(v.Input, &input); err != nil {
		t.Fatalf("parse input: %v", err)
	}
	var expected struct {
		PreimageHex  string `json:"preimage_hex"`
		EntryHashHex string `json:"entry_hash_hex"`
	}
	if err := json.Unmarshal(v.Expected, &expected); err != nil {
		t.Fatalf("parse expected: %v", err)
	}

	tbs := mustDecodeHex(t, input.TBSHex)

	// Verify preimage = 0x00 || tbs.
	preimage := append([]byte{0x00}, tbs...)
	gotPreimageHex := hex.EncodeToString(preimage)
	if gotPreimageHex != expected.PreimageHex {
		t.Errorf("preimage mismatch:\ngot  %s\nwant %s", gotPreimageHex, expected.PreimageHex)
	}

	// Verify entry hash.
	h := sha256.Sum256(preimage)
	gotHashHex := hex.EncodeToString(h[:])
	if gotHashHex != expected.EntryHashHex {
		t.Errorf("entry_hash mismatch:\ngot  %s\nwant %s", gotHashHex, expected.EntryHashHex)
	}

	// Also verify via the EntryHash helper.
	entryHash := merkle.EntryHash(tbs)
	gotHelperHex := hex.EncodeToString(entryHash)
	if gotHelperHex != expected.EntryHashHex {
		t.Errorf("EntryHash helper mismatch:\ngot  %s\nwant %s", gotHelperHex, expected.EntryHashHex)
	}
}

// --- Negative vectors: parser and verifier rejection cases ---

// TestRejectEntryIndexZero verifies that a payload with entry_index=0 is
// decoded without error (it is structurally valid) but exposes entry_index=0,
// which verifiers MUST check and reject before any other processing.
func TestRejectEntryIndexZero(t *testing.T) {
	vs := loadVectors(t)
	v := vs["reject-entry-index-zero"]

	var input struct {
		PayloadHex string `json:"payload_hex"`
	}
	if err := json.Unmarshal(v.Input, &input); err != nil {
		t.Fatalf("parse input: %v", err)
	}

	b := mustDecodeHex(t, input.PayloadHex)
	p, err := payload.Decode(b)
	if err != nil {
		t.Fatalf("unexpected decode error (entry_index=0 is structurally valid): %v", err)
	}
	if p.EntryIndex != 0 {
		t.Fatalf("expected entry_index=0, got %d — check generator", p.EntryIndex)
	}
	// The rejection must happen at the verifier level, not the parser.
	// Confirm the field is zero so downstream verifier tests can rely on it.
}

// TestRejectTruncatedPayload verifies that payload.Decode returns an error
// when the payload is truncated mid-field (last 5 bytes removed).
func TestRejectTruncatedPayload(t *testing.T) {
	vs := loadVectors(t)
	v := vs["reject-truncated-payload"]

	var input struct {
		PayloadHex string `json:"payload_hex"`
	}
	if err := json.Unmarshal(v.Input, &input); err != nil {
		t.Fatalf("parse input: %v", err)
	}

	b := mustDecodeHex(t, input.PayloadHex)
	_, err := payload.Decode(b)
	if err == nil {
		t.Error("expected parse error for truncated payload, got nil")
	}
}

// TestRejectTamperedTBS verifies that a tampered TBS produces an entry_hash
// that does not match the inclusion proof, so VerifyInclusion returns an error.
func TestRejectTamperedTBS(t *testing.T) {
	vs := loadVectors(t)
	v := vs["reject-tampered-tbs"]

	var input struct {
		PayloadHex string `json:"payload_hex"`
		RootHex    string `json:"root_hex"`
	}
	if err := json.Unmarshal(v.Input, &input); err != nil {
		t.Fatalf("parse input: %v", err)
	}

	b := mustDecodeHex(t, input.PayloadHex)
	p, err := payload.Decode(b)
	if err != nil {
		t.Fatalf("unexpected decode error: %v", err)
	}

	root := mustDecodeHex(t, input.RootHex)
	entryHash := merkle.EntryHash(p.TBS)

	// The flat VerifyInclusion uses the combined proof.
	// For this 2-entry tree the "inner" and "outer" split is trivial:
	// there is 1 hash and InnerProofCount covers it all. We test the
	// proof path using the raw Merkle function.
	err = merkle.VerifyInclusion(
		entryHash,
		int(p.EntryIndex),
		int(p.TreeSize),
		p.ProofHashes,
		root,
	)
	if err == nil {
		t.Error("expected Merkle VerifyInclusion failure for tampered TBS, got nil")
	}
}

// TestRejectWrongSigAlg verifies that a payload's sig_alg field can be decoded
// and compared against a trust config, and that a mismatch is detectable.
// The actual rejection is enforced by the verifier; this test confirms the
// field is correctly encoded and decoded.
func TestRejectWrongSigAlg(t *testing.T) {
	vs := loadVectors(t)
	v := vs["reject-wrong-sig-alg"]

	var input struct {
		PayloadHex  string `json:"payload_hex"`
		TrustConfig struct {
			SigAlg int `json:"sig_alg"`
		} `json:"trust_config"`
	}
	if err := json.Unmarshal(v.Input, &input); err != nil {
		t.Fatalf("parse input: %v", err)
	}

	b := mustDecodeHex(t, input.PayloadHex)
	p, err := payload.Decode(b)
	if err != nil {
		t.Fatalf("unexpected decode error: %v", err)
	}

	// Payload claims ECDSA P-256 (4), trust config expects Ed25519 (6).
	if p.SigAlg == uint8(input.TrustConfig.SigAlg) {
		t.Fatal("test setup error: payload and trust config have same sig_alg, expected mismatch")
	}
	if p.SigAlg != payload.SigAlgECDSAP256 {
		t.Errorf("expected payload sig_alg=%d (ECDSA P-256), got %d", payload.SigAlgECDSAP256, p.SigAlg)
	}
	if input.TrustConfig.SigAlg != int(payload.SigAlgEd25519) {
		t.Errorf("expected trust config sig_alg=%d (Ed25519), got %d", payload.SigAlgEd25519, input.TrustConfig.SigAlg)
	}
}

func TestCascadeVectorR1(t *testing.T) {
	vs := loadVectors(t)
	v := vs["revocation-cascade-r1"]

	var input struct {
		RevokedIndices []uint64 `json:"revoked_indices"`
		ValidIndices   []uint64 `json:"valid_indices"`
	}
	if err := json.Unmarshal(v.Input, &input); err != nil {
		t.Fatalf("parse input: %v", err)
	}

	var expected struct {
		CascadeHex string `json:"cascade_hex"`
		Queries    []struct {
			Index   uint64 `json:"index"`
			Revoked bool   `json:"revoked"`
		} `json:"queries"`
	}
	if err := json.Unmarshal(v.Expected, &expected); err != nil {
		t.Fatalf("parse expected: %v", err)
	}

	c, err := cascade.Build(input.RevokedIndices, input.ValidIndices)
	if err != nil {
		t.Fatalf("Build: %v", err)
	}

	gotHex := fmt.Sprintf("%x", c.Encode())
	if gotHex != expected.CascadeHex {
		t.Errorf("cascade_hex = %s, want %s", gotHex, expected.CascadeHex)
	}
	for _, q := range expected.Queries {
		if got := c.Query(q.Index); got != q.Revoked {
			t.Errorf("Query(%d) = %v, want %v", q.Index, got, q.Revoked)
		}
	}
}

func TestCascadeVectorR2(t *testing.T) {
	vs := loadVectors(t)
	v := vs["revocation-cascade-r2"]

	var expected struct {
		CascadeHex string `json:"cascade_hex"`
		Queries    []struct {
			Index   uint64 `json:"index"`
			Revoked bool   `json:"revoked"`
		} `json:"queries"`
	}
	if err := json.Unmarshal(v.Expected, &expected); err != nil {
		t.Fatalf("parse expected: %v", err)
	}

	c, err := cascade.Build(nil, nil)
	if err != nil {
		t.Fatalf("Build empty: %v", err)
	}

	gotHex := fmt.Sprintf("%x", c.Encode())
	if gotHex != expected.CascadeHex {
		t.Errorf("R2 cascade_hex = %s, want %s", gotHex, expected.CascadeHex)
	}
	for _, q := range expected.Queries {
		if got := c.Query(q.Index); got != q.Revoked {
			t.Errorf("R2 Query(%d) = %v, want %v", q.Index, got, q.Revoked)
		}
	}
}
